### PR TITLE
API: add top-level melt function as method

### PR DESCRIPTION
**Pull request by ResidentMario**
- [x] (progress on?) #12640
   - [x] pd.melt (DataFrame) (xref #14876)

 - [x] tests added / passed
 - [x] passes ``git diff upstream/master | flake8 --diff``
 - [x] whatsnew entry

This is continued from [PR#15513](https://github.com/pandas-dev/pandas/pull/15513).

### CODE REVIEW

**Comment by ResidentMario** [comment]
How should this be tested?

**Comment by jreback** [comment]
where ``pd.melt`` is currently tests, you can add in a a test like

``test_api_method`` or something and just copy an existing tests and subtitute ``DataFrame.melt`` for ``pd.melt`` (or changing a tests is ok too)

**Comment by ResidentMario** [comment]
Copying an entire test case would cause a lot of duplication though. Since `DataFrame.melt` is supposed to have the same method signature as `pd.melt`, can't we just replace `melt` with `DataFrame.melt` in the test bodies?

**Comment by jreback** [comment]
@ResidentMario yes you can simply replace them (though I would still add a *single* function to test the api (e.g. that pd.melt works)

**Comment by jreback** [comment]
also add this to ``api.rst``. and in the ``reshaing.rst`` I would change the docs.

**Comment by ResidentMario** [comment]
And and addition to `whatsnew`, presumably. Where would that go under there?

**Comment by jreback** [comment]
you could put it in Enhancements

**Comment by ResidentMario** [comment]
gtm?

File: doc/source/whatsnew/v0.20.0.txt [hunk 1]
```diff
@@ -160,6 +160,7 @@ Other enhancements
 - ``Series.sort_index`` accepts parameters ``kind`` and ``na_position`` (:issue:`13589`, :issue:`14444`)
 
 - ``DataFrame`` has gained a ``nunique()`` method to count the distinct values over an axis (:issue:`14336`).
+- ``DataFrame`` has gained a ``melt()`` method for unpivoting from a wide to long format (:issue:`12640`).
```
Comment (jreback) [inline doc/source/whatsnew/v0.20.0.txt]: note that this is the *same* as ``pd.melt``

**Comment by jreback** [comment]
lgtm. @jorisvandenbossche

**Comment by jorisvandenbossche** [comment]
Will take a look tomorrow!

**Comment by ResidentMario** [comment]
@jorisvandenbossche bump :)

**Comment by jreback** [comment]
can you rebase, this looks good though (mainly to fix the whatsnew)

**Comment by ResidentMario** [comment]
Rebased.

File: pandas/core/frame.py [hunk 2]
```diff
@@ -105,7 +105,9 @@
     optional_by="""
         by : str or list of str
             Name or list of names which refer to the axis items.""",
-    versionadded_to_excel='')
+    versionadded_to_excel='',
+    versionadded_melt='\n.. versionadded:: 0.20.0\n',
```
Comment (jorisvandenbossche) [inline pandas/core/frame.py]: this needs extra spaces (you can check `pd.DataFrame.melt?` if you are on this branch)

File: doc/source/reshaping.rst [hunk 3]
```diff
@@ -265,7 +265,7 @@ the right thing:
 Reshaping by Melt
 -----------------
 
-The :func:`~pandas.melt` function is useful to massage a
```
Comment (jorisvandenbossche) [inline doc/source/reshaping.rst]: Can you mention here both?

File: pandas/core/frame.py [hunk 4]
```diff
@@ -4051,6 +4053,108 @@ def unstack(self, level=-1, fill_value=None):
...
+    _shared_docs['melt'] = """
+    "Unpivots" a DataFrame from wide format to long format, optionally leaving
+    identifier variables set.
+
+    This function is useful to massage a DataFrame into a format where one
+    or more columns are identifier variables (`id_vars`), while all other
+    columns, considered measured variables (`value_vars`), are "unpivoted" to
+    the row axis, leaving just two non-identifier columns, 'variable' and
+    'value'.
+
```
Comment (jorisvandenbossche) [inline pandas/core/frame.py]: This blank line can be removed

**Review by jorisvandenbossche — COMMENTED** [review]
Few doc comments

**Comment by jorisvandenbossche** [comment]
@ResidentMario Sorry my 'tomorrow' took a lot longer .. !

To be clear: I am in the meantime fully convinced of the usefulness of `melt`, and having it as method on a DataFrame as counterpart of the `pivot` method is really nice.

My main concern I have (or maybe not really concern, bur rather opportunity) is about the exact API. There have been raised some issues about this (https://github.com/pandas-dev/pandas/issues/10109, https://github.com/pandas-dev/pandas/issues/14876#issuecomment-266907956), so I only think we should take this as an opportunity to look at the current `melt` API and see if we would implement it in the same way now. 
And conclusion can well be that we just want to keep it as is.

For example, IMO it would be useful to make a comparison between the R `melt` and `gather`, see how Hadley has improved it in the new `gather` version and see if we can learn something of it.

If we don't find the time for this before 0.20 release, I am OK with merging this.

**Comment by jreback** [comment]
ok @ResidentMario pls make those corrections and we can merge. If after this (in next week or two) we can figure out a better API we can discuss, otherwise can do it in next major version.

Note that also like to incorporate (and deprecate) ``wide_to_long`` here. 

btw should also look at ``cast`` and ``reshape`` R functions for inspiration.

**Comment by jorisvandenbossche** [comment]
-1 on incoroporating ``wide_to_long``. As far as I know, it is hardly used, and has an API that is not really 'pandonic'.

File: doc/source/reshaping.rst (hunk 3 above)
Comment (ResidentMario) [inline doc/source/reshaping.rst]: Done (GitHub is not folding this review component for some reason, however).

File: pandas/core/frame.py (hunk 2 above)
Comment (ResidentMario) [inline pandas/core/frame.py]: I'm not sure what you mean by this, can you clarify?

**Comment by ResidentMario** [comment]
`stack`, `unstack`, `melt`, etcetera...agree that studying the possible advantages of data transforms in other languages would be a good thing.

**Comment by jreback** [comment]
@ResidentMario 

> stack, unstack, melt, etcetera...agree that studying the possible advantages of data transforms in other languages would be a good thing.

> I don't see this being done in the next two weeks however.

not a problem

**Comment by jreback** [comment]
thanks @ResidentMario 

I made the doc-string a tiny bit more flexible (so the examples are more clear), but great patch!

File: pandas/core/frame.py (hunk 2 above)
Comment (jorisvandenbossche) [inline pandas/core/frame.py]: If you do `pd.DataFrame.melt?` you see:

```
Signature: pd.DataFrame.melt(self, id_vars=None, value_vars=None, var_name=None, value_name='value', col_level=None)
Docstring:
    "Unpivots" a DataFrame from wide format to long format, optionally leaving
    identifier variables set.

    This function is useful to massage a DataFrame into a format where one
    or more columns are identifier variables (`id_vars`), while all other
    columns, considered measured variables (`value_vars`), are "unpivoted" to
    the row axis, leaving just two non-identifier columns, 'variable' and
    'value'.

.. versionadded:: 0.20.0


    Parameters
    ----------
    ....
```

so the versionadded is not indented as the other lines.

But I think Jeff fixed it before merging

**Pull request closed without merging**